### PR TITLE
Add remote.Puller

### PR DESCRIPTION
**Pull request by jonjohnsonjr**
This PR adds a Puller implementation and uses it in `crane ls` and `crane catalog` to stream results for each page.

### CODE REVIEW

File: pkg/v1/remote/catalog.go [hunk 1]
```diff
@@ -25,8 +25,9 @@ import (
 	"github.com/google/go-containerregistry/pkg/v1/remote/transport"
 )
 
-type catalog struct {
+type Catalogs struct {
```
Comment (imjasonh) [inline pkg/v1/remote/catalog.go:28]: Why plural?

File: pkg/v1/remote/descriptor.go [hunk 2]
```diff
@@ -107,7 +107,11 @@ func Head(ref name.Reference, options ...Option) (*v1.Descriptor, error) {
 		return nil, err
 	}
 
-	return f.headManifest(o.context, ref, allManifestMediaTypes)
+	return f.head(o.context, ref)
+}
+
+func (f *fetcher) head(ctx context.Context, ref name.Reference) (*v1.Descriptor, error) {
+	return f.headManifest(ctx, ref, allManifestMediaTypes)
 }
```
Comment (imjasonh) [inline pkg/v1/remote/descriptor.go:115]: This method doesn't add much, can we just have callers pass `allManifestMediaTypes` themselves?

**Review by imjasonh — APPROVED** [review]

File: pkg/v1/remote/catalog.go (hunk 1 above)
Comment (jonjohnsonjr) [reply pkg/v1/remote/catalog.go:28]: `remote.Catalog` is already defined 😅

File: pkg/v1/remote/descriptor.go (hunk 2 above)
Comment (jonjohnsonjr) [reply pkg/v1/remote/descriptor.go]: Done.

**Pull request merged**